components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodTerminalService.kt

```diff
@@ -0,0 +1,167 @@
+        val toolWindowManagerListener = object : ToolWindowManagerListener {
+            override fun toolWindowsRegistered(ids: MutableList<String>, toolWindowManager: ToolWindowManager) {
+                if (ids.contains(TerminalToolWindowFactory.TOOL_WINDOW_ID)) {
+                    debug("TerminalToolWindow got registered!")
+                    completableFuture.complete(null)
+                    messageBusConnection.disconnect()
+                }
+            }
+        }
+
+        messageBusConnection.subscribe(ToolWindowManagerListener.TOPIC, toolWindowManagerListener)
+
+        return completableFuture
+    }
+
+    private fun getSupervisorTasksList(): CompletableFuture<List<TaskStatus>> {
+        val completableFuture = CompletableFuture<List<TaskStatus>>()
+        val taskStatusRequest = TasksStatusRequest.newBuilder().setObserve(true).build()
+        val taskStatusResponseObserver = object : ClientResponseObserver<TasksStatusRequest, TasksStatusResponse> {
+            override fun beforeStart(request: ClientCallStreamObserver<TasksStatusRequest>) {
+                lifetime.onTerminationOrNow {
+                    request.cancel(null, null)
+                }
+            }
+
+            override fun onNext(response: TasksStatusResponse) {
+                debug("Received task list: ${response.tasksList}")
+
+                for (task in response.tasksList) {
+                    if (task.state === TaskState.opening) {
+                        return
+                    }
+                }
+
+                this.onCompleted()
+                completableFuture.complete(response.tasksList)
+            }
+
+            override fun onCompleted() {
+                debug("Successfully fetched tasks from Supervisor.")
+            }
+
+            override fun onError(throwable: Throwable) {
+                thisLogger().error("Got an error while trying to fetch tasks from Supervisor.", throwable)
+                completableFuture.completeExceptionally(throwable)
+            }
+        }
+
+        statusServiceStub.tasksStatus(taskStatusRequest, taskStatusResponseObserver)
+
+        return completableFuture
+    }
+
+    private fun getSupervisorTerminalsListAsync(): Deferred<TerminalOuterClass.ListTerminalsResponse> {
+        val listTerminalsRequest = TerminalOuterClass.ListTerminalsRequest.newBuilder().build()
+        return terminalServiceFutureStub.list(listTerminalsRequest).asDeferred()
+    }
+
+    private fun createSharedTerminal(supervisorTerminal: TerminalOuterClass.Terminal) {
+        debug("Creating shared terminal '${supervisorTerminal.title}' on Backend IDE")
+
+        val shellTerminalWidget = terminalView.createLocalShellWidget(project.basePath, supervisorTerminal.title)
+
+        backendTerminalManager.shareTerminal(shellTerminalWidget, supervisorTerminal.alias)
+
+        shellTerminalWidget.executeCommand("gp tasks attach ${supervisorTerminal.alias}")
```

Comment:
we are loosing control with that, no? We need to propagate proper UI sizes to terminal, close it then a user closes, and close a frontend terminal then an underlying terminal is closed?

Reply:
Indeed. But I can't get back using CloudTerminalRunner until [#CWM-5523: Support CloudTerminalRunner mirroring to thin client](https://youtrack.jetbrains.com/issue/CWM-5523) is resolved. ([Internal Discussion](https://gitpod.slack.com/archives/C02BRJLGPGF/p1646752507164589?thread_ts=1646056923.510109&cid=C02BRJLGPGF))

There are like 5x more issues when using CloutTerminalRunner in place of `go tasks attach`.

Maybe the best solution for this changes is creating a custom IntelliJ Tool Window, as suggested by JetBrains Team, but that would require a totally different implementation/PR.

Reply:
> Maybe the best solution for this changes is creating a custom IntelliJ Tool Window, as suggested by JetBrains Team, but that would require a totally different implementation/PR.

Agree it is better to wrap up what we have and tackle other things in followed up.

It should at least open all task terminals, not like in https://github.com/gitpod-io/gitpod/pull/8338#issuecomment-1098878220. Fixing closing behaviour will be good but let's do it with another PR too.

Reply:
I'm now trying to use `backendTerminalManager.createNewSharedTerminal` (as when I use it, all the terminals are created accordingly and displayed on the Thin Client), but the consequence is that there's no easy way to find which terminal was created by this method.
I'm looking into a way to find each terminal and run `gp attach` in each. (https://github.com/gitpod-io/gitpod/pull/8338/commits/b82c26ed84fb665d6918392c3f35e15b5fe0c2a4)